### PR TITLE
feat: unified privacy model — Visibility API, sync leak fix, workspace access policies

### PR DESCRIPTION
**Pull request by Jurij89**
## Summary

This PR unifies the three separate privacy mechanisms in DKG-V9 (paranet `private` flag, workspace `localOnly` flag, publish-time `accessPolicy`) into a single, consistent `Visibility` type used across all operations. It also fixes a **critical data leak** where the sync protocol served workspace data to any peer regardless of privacy settings.

### Problem

Privacy in DKG-V9 was configured through three independent mechanisms with different APIs:

| Mechanism | Flag | Actual Protection |
|---|---|---|
| Paranet creation | `private: true` | All-or-nothing isolation (no sharing with specific peers) |
| Workspace writes | `localOnly: true` | **False privacy**: only suppressed broadcast, sync still served data to anyone |
| Published data | `accessPolicy: 'ownerOnly' \| 'allowList'` | Real enforcement via Ed25519 signatures |

Key issues:
- **Sync protocol data leak**: `localOnly` workspace data was served to any peer via `/dkg/sync/1.0.0`
- **No workspace access policies**: workspace data was either broadcast to all or local-only, with no per-peer control
- **Inconsistent APIs**: three different parameter names (`private`, `localOnly`, `accessPolicy`) for the same concept
- **Unsafe defaults**: workspace writes broadcast to all subscribers by default

#### New unified `Visibility` type
```typescript
type Visibility =
  | 'private'           // Only this node
  | 'public'            // Anyone can read
  | { peers: string[] } // Only listed peer IDs

// Used consistently on ALL operations:
await agent.createParanet({ id: 'x', name: 'X', visibility: 'private' });
await agent.writeToWorkspace('x', quads, { visibility: { peers: ['12D3KooW...'] } });
await agent.publish('x', quads, { visibility: 'public' });
await agent.enshrineFromWorkspace('x', 'all', { visibility: { peers: ['12D3KooW...'] } });
```

#### `resolveVisibility()` helper
Maps the new `Visibility` type to internal `AccessPolicy`/`allowedPeers`/`broadcast` with full backward compatibility for legacy parameters.

#### Critical sync fix
The `PROTOCOL_SYNC` handler now:
1. Accepts the requesting peer's ID
2. Queries `_workspace_meta` for per-operation access policies
3. Filters workspace responses — only serves data the requesting peer is authorized to see
4. Treats pre-migration data (no access policy triple) as public for backward compatibility

#### Workspace access policy plumbing
- Protobuf: new fields `accessPolicy` (field 9) and `allowedPeers` (field 10) on `WorkspacePublishRequest` — additive, backward compatible
- Metadata: `generateWorkspaceMetadata()` emits `dkg:accessPolicy` and `dkg:allowedPeer` triples
- Publisher: threads access policy through workspace write pipeline
- Handler: extracts and stores access policy from incoming gossip messages

#### Default behavior changes
| Operation | Old Default | New Default | Rationale |
|---|---|---|---|
| `writeToWorkspace()` | Broadcast to all (`localOnly: false`) | Private (`visibility: 'private'`) | Drafts should not leak |
| `createParanet()` | Public | Public (unchanged) | Paranets are collaboration containers |
| `enshrineFromWorkspace()` | Public (no option) | Public (now configurable) | Enshrinement is explicit "share" |
| `publish()` | Context-dependent | Uses `resolveVisibility()` | Consistent resolution |

#### Full backward compatibility
All legacy parameters continue to work:
- `localOnly: true` → maps to `visibility: 'private'`
- `private: true` → maps to `visibility: 'private'`
- `accessPolicy: 'ownerOnly'` → maps to `visibility: 'private'`
- `accessPolicy: 'allowList', allowedPeers: [...]` → maps to `visibility: { peers: [...] }`

When both `visibility` and legacy params are provided, `visibility` takes precedence.

## Changes by commit

1. **`2a9285e` feat: add AccessPolicy, Visibility types and resolveVisibility() to core**
   - New types in `packages/core/src/types.ts`
   - New `resolveVisibility()` in `packages/core/src/visibility.ts`
   - `access-handler.ts` imports AccessPolicy from core (re-exports for compat)

2. **`c8d06b1` fix: filter workspace data by access policy in PROTOCOL_SYNC handler**
   - Sync handler accepts `peerId`, queries workspace_meta, filters by policy
   - Pre-migration data without policy triples treated as public

3. **`90aa3d9` feat: thread accessPolicy and allowedPeers through workspace publish pipeline**
   - Protobuf fields 9/10 added
   - Metadata generation emits access policy triples
   - Publisher and handler thread the new fields

4. **`9304b76` feat: unify API surface with Visibility parameter across all agent methods**
   - `writeToWorkspace`, `writeConditionalToWorkspace`, `publish`, `createParanet`, `enshrineFromWorkspace` all accept `visibility`
   - Legacy params deprecated but functional

5. **`aa332f2` fix: pass publisherPeerId through enshrineFromWorkspace to publish**
   - Bug found by QA: enshrine with non-public policy failed because publisherPeerId wasn't forwarded

6. **`39e6c26` test: comprehensive QA for unified privacy model — 69 new tests**
   - 6 new test files across core, publisher, and agent packages
   - Covers: resolveVisibility, protobuf round-trip, metadata emission, sync filtering, API defaults, backward compat

## Files changed

**Core (foundation):**
- `packages/core/src/types.ts` — `AccessPolicy`, `Visibility` types
- `packages/core/src/visibility.ts` — `resolveVisibility()` helper
- `packages/core/src/index.ts` — Exports
- `packages/core/src/proto/workspace.ts` — Protobuf fields 9/10

**Publisher (plumbing):**
- `packages/publisher/src/access-handler.ts` — Import from core
- `packages/publisher/src/dkg-publisher.ts` — Workspace + enshrine options
- `packages/publisher/src/metadata.ts` — Access policy quads in workspace metadata
- `packages/publisher/src/workspace-handler.ts` — Extract policy from gossip

**Agent (API + enforcement):**
- `packages/agent/src/dkg-agent.ts` — Sync filter, visibility on all methods

**Tests (69 new):**
- `packages/core/test/visibility.test.ts` (20 tests)
- `packages/core/test/workspace-proto-privacy.test.ts` (8 tests)
- `packages/publisher/test/workspace-metadata-privacy.test.ts` (11 tests)
- `packages/publisher/test/workspace-privacy.test.ts` (10 tests)
- `packages/agent/test/visibility-api.test.ts` (15 tests)
- `packages/agent/test/sync-access-filtering.test.ts` (5 tests)

## Test plan

- [x] `resolveVisibility()` — all visibility/legacy combos, precedence, defaults, edge cases (20 tests)
- [x] Protobuf round-trip — fields 9/10 encode/decode, backward compat with old messages (8 tests)
- [x] Workspace metadata — access policy quad emission for all policy types (11 tests)
- [x] Publisher pipeline — workspace write + handler + enshrine with access policies (10 tests)
- [x] Agent API — broadcast control, paranet policy storage, default changes, legacy compat (15 tests)
- [x] Sync filtering — mixed-visibility workspace data filtered correctly by peer (5 tests)
- [x] All existing tests pass (core: 198, publisher: 335, agent: 129)
- [ ] Manual testing: create paranet with `visibility: { peers: [...] }`, write to workspace, verify sync filters correctly
- [ ] Manual testing: verify legacy `localOnly: true` and `private: true` still work identically

## Out of scope (future work)

- Gossip message encryption for non-public workspace broadcasts (residual eavesdropping risk)
- GossipSub topic validators for paranet-level membership
- Revocation mechanism for access lists
- New OpenClaw agent tools (`dkg_workspace_write`, `dkg_publish_from_workspace`)
- Removal of deprecated legacy parameters (next major version)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/agent/src/dkg-agent.ts [hunk 1]
```diff
@@ -1254,17 +1355,38 @@ export class DKGAgent {
...
+    opts?: {
+      visibility?: Visibility;
+      /** @deprecated Use visibility: 'private' instead */
+      localOnly?: boolean;
+      operationCtx?: OperationContext;
+    },
+  ): Promise<{ workspaceOperationId: string }> {
+    // When neither visibility nor localOnly is specified, default to private
+    const effectiveVisibility = opts?.visibility ?? (opts?.localOnly === false ? 'public' : undefined);
+    const resolved = resolveVisibility(effectiveVisibility, { localOnly: opts?.localOnly ?? true });
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts]: Fixed in `5f4b1d9`. Reverted to the old broadcast-by-default behavior when neither `visibility` nor `localOnly` is specified. When `opts` is omitted or both fields are absent, `resolveVisibility()` now falls through to the public/broadcast default, preserving backward compatibility. The same fix was applied to `writeConditionalToWorkspace()`. The safe-default (private) will be introduced in a future major version with a coordinated migration.

File: packages/agent/src/dkg-agent.ts [hunk 2]
```diff
@@ -1254,17 +1355,38 @@ export class DKGAgent {
...
-    this.log.info(ctx, `Writing ${quads.length} quads to workspace for paranet ${paranetId}${opts?.localOnly ? ' (local-only)' : ''}`);
+    this.log.info(ctx, `Writing ${quads.length} quads to workspace for paranet ${paranetId} (visibility=${typeof effectiveVisibility === 'object' ? 'peers' : effectiveVisibility ?? 'private'})`);
     const { workspaceOperationId, message } = await this.publisher.writeToWorkspace(paranetId, quads, {
       publisherPeerId: this.node.peerId.toString(),
       operationCtx: ctx,
+      accessPolicy: resolved.accessPolicy,
+      allowedPeers: resolved.allowedPeers,
     });
-    if (!opts?.localOnly) {
+    if (resolved.broadcast) {
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts]: Fixed in `5f4b1d9`. `allowList` workspace writes no longer broadcast on GossipSub. Two changes: (1) `resolveVisibility()` now returns `broadcast: false` for allowList, (2) the agent broadcast gate was changed to `resolved.accessPolicy === "public"` as defense-in-depth. AllowList workspace data now reaches peers only via the sync protocol, which enforces the per-operation access filter. Gossip encryption for allowList is tracked as future work.

File: packages/agent/src/dkg-agent.ts [hunk 3]
```diff
@@ -468,6 +563,8 @@ export class DKGAgent {
           const metaResult = await this.store.query(metaQuery);
           if (metaResult.type === 'bindings') {
             for (const b of metaResult.bindings) {
+              // Filter metadata: only include triples whose subject is an accessible operation
+              if (accessibleOps.size > 0 && !accessibleOps.has(b['s'])) continue;
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts]: Fixed in `5f4b1d9`. The sync filter now has a two-part inclusion check: triples are included if (a) the subject is an accessible operation URI (original behavior), OR (b) the predicate is `dkg:workspaceOwner` and the subject is an accessible root entity. This ensures `reconstructWorkspaceOwnership()` rebuilds correctly on synced replicas after restart.

File: packages/agent/src/dkg-agent.ts [hunk 4]
```diff
@@ -1254,17 +1363,43 @@ export class DKGAgent {
...
       publisherPeerId: this.node.peerId.toString(),
       operationCtx: ctx,
+      accessPolicy: resolved.accessPolicy,
+      allowedPeers: resolved.allowedPeers,
     });
-    if (!opts?.localOnly) {
+    // Only broadcast on GossipSub for public data. allowList data must NOT be
+    // gossipped because any subscribed peer can read raw gossip messages before
+    // the sync-side filter runs. allowList data reaches peers via sync only.
+    if (resolved.accessPolicy === 'public') {
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts:1437]: Fixed in `b709d62`. Added `PROTOCOL_WORKSPACE_PUSH` (`/dkg/workspace-push/1.0.0`) for targeted delivery. After an allowList workspace write, `pushWorkspaceToAllowedPeers()` iterates currently-connected peers, filters to those in the allowed list, and sends the workspace message directly via the stream protocol. Handler on the receiving side delegates to the existing `WorkspaceHandler.handle()`. Applied to both `writeToWorkspace` and `writeConditionalToWorkspace`.

File: packages/agent/src/dkg-agent.ts [hunk 5]
```diff
@@ -452,13 +537,23 @@ export class DKGAgent {
...
+            for (const row of accessResult.bindings) {
+              const re = row['re'];
+              const op = row['op'];
+              if (re && op && accessibleRoots.has(re)) accessibleOps.add(op);
+            }
+          }
+
           const metaQuery = cutoff != null
             ? `SELECT ?s ?p ?o WHERE {
                 GRAPH <${wsMetaGraph}> { ?s ?p ?o }
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts]: Fixed in `b709d62`. Rewrote the workspace metadata sync query to use UNION: first branch fetches operation triples with non-expired `publishedAt`, second branch fetches `dkg:workspaceOwner` triples via BIND. Post-query access control filter still applies to both — owner triples are only included for accessible root entities.

File: packages/core/src/visibility.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    if (visibility === 'private') {
+      return { accessPolicy: 'ownerOnly', allowedPeers: [], broadcast: false };
+    }
+    if (visibility === 'public') {
+      return { accessPolicy: 'public', allowedPeers: [], broadcast: true };
+    }
+    if (typeof visibility === 'object' && 'peers' in visibility) {
+      // allowList must NOT broadcast on GossipSub — any subscribed peer can
+      // read raw gossip messages. allowList data reaches peers via sync only.
+      return { accessPolicy: 'allowList', allowedPeers: visibility.peers, broadcast: false };
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/core/src/visibility.ts]: Fixed in `b709d62`. `resolveVisibility()` now trims whitespace, filters empty strings, deduplicates via Set, and throws if the cleaned list is empty: `"visibility { peers: [...] } requires at least one valid peer ID"`. Added 6 new tests covering edge cases (whitespace, duplicates, all-empty, mixed entries).

File: packages/agent/src/dkg-agent.ts [hunk 7]
```diff
@@ -1254,23 +1386,54 @@ export class DKGAgent {
...
+      visibility?: Visibility;
+      /** @deprecated Use visibility: 'private' instead */
+      localOnly?: boolean;
+      operationCtx?: OperationContext;
+    },
+  ): Promise<{ workspaceOperationId: string }> {
+    // Preserve backward compatibility: when neither visibility nor localOnly is
+    // specified, default to public/broadcast (matching pre-migration behavior).
+    // Only go private when the caller explicitly requests it.
+    const effectiveVisibility = opts?.visibility ?? (opts?.localOnly != null ? undefined : 'public');
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts]: Fixed in 9e086ce. Inner operations now inherit the paranet stored visibility as default. Added getParanetDefaultVisibility(paranetId) helper that reads from the in-memory subscribedParanets map. writeToWorkspace, writeConditionalToWorkspace, and publish all call this when no explicit visibility is passed. Falls back to public for pre-existing paranets with no stored policy. enshrineFromWorkspace keeps its public default but logs a warning when the paranet has a more restrictive policy.

File: packages/publisher/src/dkg-publisher.ts [hunk 8]
```diff
@@ -249,6 +253,8 @@ export class DKGPublisher implements Publisher {
       timestampMs: Date.now(),
       operationId: ctx.operationId,
       casConditions,
+      accessPolicy: options.accessPolicy,
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/publisher/src/dkg-publisher.ts:275]: Fixed in 9e086ce. Added validation in _writeToWorkspaceImpl() matching the same invariant checks as publish(): rejects unknown accessPolicy values, requires non-empty allowedPeers for allowList, and strips contradictory allowedPeers on ownerOnly (with warning log).

File: packages/agent/src/dkg-agent.ts [hunk 9]
```diff
@@ -1749,12 +1995,12 @@ export class DKGAgent {
...
     this.subscribedParanets.set(opts.id, {
       name: opts.name,
-      subscribed: !opts.private,
+      subscribed: !isPrivate,
       synced: true,
       onChainId: onChainId,
     });
 
-    if (!opts.private) {
+    if (!isPrivate) {
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts]: Fixed in 9e086ce. Replaced single isPrivate gate with two: skipChain (ownerOnly only) and skipDefinitionBroadcast (all non-public). allowList paranets now register on-chain and subscribe to paranet-specific GossipSub topics, but do NOT broadcast their definition on the public ontology topic. Discovery for allowed peers happens through direct communication, not public gossip.

File: packages/agent/src/dkg-agent.ts [hunk 10]
```diff
@@ -1180,13 +1312,19 @@ export class DKGAgent {
         );
       }
     }
+    // Inherit paranet default visibility when no explicit visibility/accessPolicy is given
+    const explicitVisibility = opts?.visibility ?? (opts?.accessPolicy ? undefined : this.getParanetDefaultVisibility(paranetId));
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts:1321]: Not a new bug — this is the pre-existing publish design. The `accessPolicy` field has always only controlled access to `privateQuads`, not `quads` (public triples). Public triples are public by design regardless of access policy. The visibility inheritance sets a safer default for private-triple access control, but does not change public-triple behavior. If users want ALL data protected, they should put it in `privateQuads`. Will add a clarifying code comment to prevent confusion.

File: packages/agent/src/dkg-agent.ts [hunk 11]
```diff
@@ -1279,40 +1449,109 @@ export class DKGAgent {
...
+    );
+  }
+
+  /**
+   * Look up the paranet's stored access policy from the in-memory subscribedParanets map.
+   * Returns undefined when no policy is stored (i.e. the paranet is public or pre-dates the
+   * privacy model). Callers use this to inherit the paranet's default when the user omits
+   * an explicit visibility parameter on a write/publish operation.
+   */
+  private getParanetDefaultVisibility(paranetId: string): Visibility | undefined {
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts:1541]: Fixed in 328a401. Three changes: (1) discoverParanetsFromStore() now queries for dkg:accessPolicy and dkg:allowedPeer triples and populates the subscription map on restart. (2) discoverParanetsFromChain() maps on-chain accessPolicy number (0=public, 1=permissioned) to the AccessPolicy type. (3) getParanetDefaultVisibility() has a store-backed fallback — if the in-memory map has no policy, it queries the paranet meta graph before returning undefined.

File: packages/agent/src/dkg-agent.ts [hunk 12]
```diff
@@ -1749,15 +2035,23 @@ export class DKGAgent {
...
       synced: true,
       onChainId: onChainId,
+      accessPolicy: resolved.accessPolicy !== 'public' ? resolved.accessPolicy : undefined,
+      allowedPeers: resolved.allowedPeers.length > 0 ? resolved.allowedPeers : undefined,
     });
 
-    if (!opts.private) {
-      // Auto-subscribe to the new paranet's GossipSub topic
+    // ownerOnly: fully isolated — no gossip topics, no broadcast
+    // allowList: subscribe to paranet-specific topics (so allowed peers can gossip
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts:2166]: Fixed in 328a401. Non-public paranet definitions are now stored in the paranet own meta graph (paranetMetaGraphUri) instead of the shared ontology graph. Public paranets continue using the ontology graph (unchanged). discoverParanetsFromStore() also filters out any non-public paranets that may have leaked into the ontology graph from pre-fix code, preventing auto-discovery. 6 new tests cover all scenarios.

File: packages/agent/src/dkg-agent.ts [hunk 13]
```diff
@@ -422,25 +430,106 @@ export class DKGAgent {
...
         // we return the first N non-expired triples. Only include exact root subject
         // or skolemized children (/.well-known/genid/...) to avoid pulling unrelated
         // entities that share a URI prefix (e.g. urn:x vs urn:x/other).
+        // Additionally filter by accessible root entities for access control.
         const cutoff = wsTtl > 0 ? new Date(Date.now() - wsTtl).toISOString() : null;
+        const rootValues = [...accessibleRoots].map(r => `<${r}>`).join(' ');
+        const hasAccessFilter = accessibleRoots.size > 0;
+        const accessFilterClause = hasAccessFilter ? `VALUES ?re { ${rootValues} }` : '';
+
         const wsQuery =
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts:509]: This is the pre-existing 'no revocation' limitation documented in both the privacy report (Gap 2) and the PR description (Out of scope section). Implementing workspace data revocation requires a tombstone/delete protocol and is tracked as future work. Not a regression from this PR.

File: packages/agent/src/dkg-agent.ts [hunk 14]
```diff
@@ -1279,40 +1454,172 @@ export class DKGAgent {
...
+   * Returns undefined when no policy is stored (i.e. the paranet is public or pre-dates the
+   * privacy model). Callers use this to inherit the paranet's default when the user omits
+   * an explicit visibility parameter on a write/publish operation.
+   *
+   * If the in-memory map has no accessPolicy, falls back to a one-time store query
+   * (checking both the ontology graph and the paranet's _meta graph) and caches the result.
+   */
+  private getParanetDefaultVisibility(paranetId: string): Visibility | undefined {
+    const sub = this.subscribedParanets.get(paranetId);
+    if (!sub) return undefined;
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts]: Fixed in c92b7cd. Added rebuildNonPublicParanets() method that scans all _meta graphs for dkg:accessPolicy triples on startup and rebuilds subscription entries. Also fixed getParanetDefaultVisibility() to trigger the store-backed fallback even when the subscription entry is entirely absent, not just when it exists but lacks the policy field.

File: packages/agent/src/dkg-agent.ts [hunk 15]
```diff
@@ -2169,6 +2567,60 @@ export class DKGAgent {
     if (discovered > 0) {
       this.log.info(ctx, `Auto-subscribed to ${discovered} new paranet(s) from store`);
     }
+
+    // Rehydrate access policy for paranets already in the subscription registry
+    // whose accessPolicy/allowedPeers are missing (e.g. after restart).
+    // Check both the ontology graph (public) and the paranet's own _meta graph (non-public).
+    for (const [id, sub] of this.subscribedParanets) {
+      if (sub.accessPolicy !== undefined) continue; // already populated
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/src/dkg-agent.ts]: Fixed in c92b7cd. Changed the rehydration guard to also trigger when accessPolicy is allowList but allowedPeers is absent or empty. Chain-discovered entries (which set the policy from the on-chain number but lack peer lists) are now completed from _meta graph data.

File: packages/agent/test/visibility-api.test.ts [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+      id: paranet,
+      name: 'Private With Policy',
+      visibility: 'private',
+    });
+
+    const DKG_NS = 'http://dkg.io/ontology/';
+    const ontologyGraph = 'did:dkg:paranet:ontology';
+    const paranetUri = `did:dkg:paranet:${paranet}`;
+    const result = await agent.query(
+      `SELECT ?policy WHERE { GRAPH <${ontologyGraph}> { <${paranetUri}> <${DKG_NS}accessPolicy> ?policy } }`,
```
[the remark this answers is not in the document]
Comment (Jurij89) [reply packages/agent/test/visibility-api.test.ts]: Fixed in c92b7cd. Updated test assertions to query paranetMetaGraphUri() instead of the shared ontology graph for non-public paranet definitions.
